connpicker.go

```diff
@@ -0,0 +1,111 @@
+package gocql
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+)
+
+type ConnPicker interface {
+	Pick(token) *Conn
+	Put(*Conn)
+	Remove(conn *Conn)
+	Size() (int, int)
+	Close()
+}
+
+type defaultConnPicker struct {
+	conns []*Conn
+	pos   uint32
+	size  int
+	mu    sync.RWMutex
+}
+
+func newDefaultConnPicker(size int) *defaultConnPicker {
+	if size <= 0 {
+		panic(fmt.Sprintf("invalid pool size %d", size))
+	}
+	return &defaultConnPicker{
+		size: size,
+	}
+}
+
+func (p *defaultConnPicker) Remove(conn *Conn) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	for i, candidate := range p.conns {
+		if candidate == conn {
+			p.conns[i] = nil
+			return
+		}
+	}
+}
+
+func (p *defaultConnPicker) Close() {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	conns := p.conns
+	p.conns = nil
+	for _, conn := range conns {
+		if conn != nil {
+			conn.Close()
+		}
+	}
+}
+
+func (p *defaultConnPicker) Size() (int, int) {
+	size := len(p.conns)
+	return size, p.size - size
+}
+
+func (p *defaultConnPicker) Pick(token) *Conn {
+	pos := int(atomic.AddUint32(&p.pos, 1) - 1)
+	size := len(p.conns)
+
+	var (
+		leastBusyConn    *Conn
+		streamsAvailable int
+	)
+
+	// find the conn which has the most available streams, this is racy
+	for i := 0; i < size; i++ {
+		conn := p.conns[(pos+i)%size]
+		if conn == nil {
+			continue
+		}
+		if streams := conn.AvailableStreams(); streams > streamsAvailable {
+			leastBusyConn = conn
+			streamsAvailable = streams
+		}
+	}
+
+	return leastBusyConn
+}
+
+func (p *defaultConnPicker) Put(conn *Conn) {
+	p.mu.Lock()
+	p.conns = append(p.conns, conn)
+	p.mu.Unlock()
+}
+
+// nopConnPicker is a no-operation implementation of ConnPicker.
+type nopConnPicker struct{}
+
+func (nopConnPicker) Pick(token) *Conn {
+	return nil
+}
+
+func (nopConnPicker) Put(*Conn) {
+}
+
+func (nopConnPicker) Remove(conn *Conn) {
+}
+
+func (nopConnPicker) Size() (int, int) {
+	return 0, 1 // 1 makes hostConnPool want to fill the pool
```

Comment:
is there a reason to _want_ to fill the pool for no-op?

Reply:
nopConnPicker is a *guard* implementation used in newHostConnPool [1]. If it's 0 then hostConnPool.fill is not called. The real ConnPool assignment is deferred to the moment when we have a connection [2]. Maybe we should change the comment, do you have any suggestions?

[1] https://github.com/gocql/gocql/pull/1210/files#diff-3aa88c8ac2c0ffb49cbd55573005b354R283
[2] https://github.com/gocql/gocql/pull/1210/files#diff-3aa88c8ac2c0ffb49cbd55573005b354R515

Reply:
I'm having a couple of ideas here, will try something out first.

Reply:
ok, I'm fine with this guard. Maybe update the comment that if hostConnPool doesn't need to fill the pool, then it won't initiate the picker in this case?

Reply:
I added more detailed description, PTAL.